hpx/lcos/spmd_block.hpp

```diff
@@ -0,0 +1,211 @@
+//  Copyright (c) 2017 Antoine Tran Tan
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#if !defined(HPX_LCOS_SPMD_BLOCK_HPP)
+#define HPX_LCOS_SPMD_BLOCK_HPP
+
+#include <hpx/include/plain_actions.hpp>
+#include <hpx/lcos/future.hpp>
+#include <hpx/lcos/barrier.hpp>
+#include <hpx/lcos/broadcast.hpp>
+#include <hpx/parallel/execution_policy.hpp>
+#include <hpx/runtime/get_locality_id.hpp>
+#include <hpx/runtime/launch_policy.hpp>
+#include <hpx/runtime/serialization/serialize.hpp>
+#include <hpx/traits/concepts.hpp>
+#include <hpx/traits/is_action.hpp>
+
+#include <boost/range/irange.hpp>
+
+#include <cstddef>
+#include <functional>
+#include <memory>
+#include <string>
+#include <type_traits>
+#include <utility>
+
+namespace hpx { namespace lcos
+{
+    namespace detail
+    {
+        template <typename T>
+        struct extract_first_parameter
+        {};
+
+        template <>
+        struct extract_first_parameter< hpx::util::tuple<> >
+        {
+            using type = std::false_type;
+        };
+
+        template <typename Arg0, typename... Args>
+        struct extract_first_parameter< hpx::util::tuple<Arg0, Args...> >
+        {
+            using type = typename std::decay<Arg0>::type;
+        };
+    }
```

Comment:
This should be factored out and merged with the one defined in `hpx/lcos/local/spmd_block.hpp`